### PR TITLE
Experimental support for user extensible cppcheck  options

### LINKED ISSUE #22: Support for user extended options (other_options)
Per kind request in recent PR #20 discussion:
>Let me ask you something, would you please take some time to add a little feature to it once I merge this?
>
>Let users specify custom options ( without checking them inside my code). As many >options are added to the cppcheck tool on every new release, It's not possible to cover >all of them. so I was planning to add an environment variable like `other_options` where >users can add options on their own.

_Originally posted by @deep5050 in https://github.com/deep5050/cppcheck-action/issues/20#issuecomment-751683891_

### PR DESCRIPTION
**Pull request by sthagen**
This change set implements the processing logic to inject user (unchecked) options into the cppcheck commandline call vector.

*Note*: There is yet no documentation informing the users that they can use this.

Example:
```other_options: --report-progress --xml -v -l 2```

... not sure about the yaml syntax here, but in commandline environment the following setting works:
```OTHER_OPTIONS="--report-progress --xml -v -l 2    " python ./entrypoint.py```

This resolves #22 (the implementation part).

Any feedback much appreciated.

PS: The version and help calls were reset to not check return code. This is considered good, as it is not assumed we really get the name of the binary called wrong ... and simplifies error handling.

### CODE REVIEW

File: src/entrypoint.py [hunk 1]
```diff
@@ -72,6 +74,7 @@
 MAX_CTU_DEPTH = "INPUT_MAX_CTU_DEPTH"
 OUTPUT_FILE = "INPUT_OUTPUT_FILE"
 PLATFORM_TYPE = "INPUT_PLATFORM"
+OTHER_OPTIONS = "OTHER_OPTIONS"
```
Comment (deep5050) [inline src/entrypoint.py:77]: OTHER_OPTIONS = "INPUT_OTHER_OPTIONS"

**Review by deep5050 — CHANGES_REQUESTED** [review]
Also replace the action.yaml with

```yaml
name: "cppcheck-action"
author: deep5050
description: "Find potential vulnerabilities in your C/C++ codes"

inputs:
  github_token:
    description: "GITHUB_TOKEN secret < automatically provided by Github , you don't need to set up >"
    required: true
    
  # github_username:
  #   description: GitHub account name
  #   default: "cppcheck-action"
  # github_email:
  #   description: GitHub account email
  #   default: "cppcheck-action@master"
  # commit_msg:
  #   description: custom commit message
  #   default: "cppcheck report added or updated"

  check_library:
    description: "Show information messages when library files have info."
    default: "disable"
    required: false

  skip_preprocessor:
    description: "Print preprocessor output on stdout and don't do any further processing."
    default: "disable"
    required: false

  enable:
    description: "Enable additional checks"
    default: "all"
    required: false

  exclude_check:
    description: "Give a source file or source file directory to exclude from the check"
    default: "disable"
    required: false

  inconclusive:
    description: " Allow that Cppcheck reports even though the analysis is inconclusive."
    default: "enable"
    required: false

  inline_suppression:
    description: "Enable inline suppressions"
    default: "disable"
    required: false

  force_language:
    description: "Forces cppcheck to check all files as the given language"
    default: "disable"
    required: false

  max_ctu_depth:
    description: "Max depth in whole program analysis"
    default: "disable"
    required: false

  platform:
    description: "Specifies platform specific types and sizes"
    default: "disable"
    required: false

  output_file:
    description: " files where the result to be dumped"
    default: "cppcheck_report.txt"
    required: false
  target_branch:
    description: "Branch that the badge will target. Defaults to the current branch."
    default: ""
    required: false
    
  other_options:
    descriptoin: "Any other options you want to add ,seperate with a space, wrong options will cause a failure"
    default: "disable"
    required: false
   

  pull_request_branch:
    description: "Branch that the action will target on pull_request. Defaults to GITHUB_BASE_REF."
    default: ""
    required: false

  target_repository:
    description: "Repository that the report will be pushed on. Defaults to the current repository."
    default: ""
    required: false

  pull_request_repository:
    description: "Repository that the action will target on pull_request. Defaults to TARGET_REPOSITORY"
    default: ""
    required: false

runs:
  using: "docker"
  image: "Dockerfile"

branding:
  icon: "check-circle"
  color: "red"

```

**Comment by sthagen** [comment]
> Also replace the action.yaml with
> 
> ```yaml
> ...
>   other_options:
>     description: "Any other options you want to add ,seperate with a space, wrong options will cause a failure"
> ```

Small typo `descriptoin` -> `description` fixed.

File: src/entrypoint.py (hunk 1 above)
Comment (sthagen) [reply src/entrypoint.py]: Done (and amended per below). Thanks for spotting and adding the documentation / configuration via the YAML file.

**Comment by deep5050** [comment]
Looks good. 
I'll test it in a branch and let you know ASAP.

**Comment by deep5050** [comment]
@sthagen  please enable discussion on your repo. we need to discuss something there :1:

**Comment by deep5050** [comment]
@all-contributors please add @sthagen  for infrastructure, tests and code

**Comment by BaderEddineOuaich** [comment]
Hello @sthagen, i hope you're doing fine.
i hope we revive back this pr because its really useful!
i have tested the other_options addition you made in another repo (the only way i could test this without docker)
i have created 2 repositories:
- the first is a fork of this repository plus this pull request additions: https://github.com/BaderEddineOuaich/cppcheck-action
- the second is a new repository which tests the additions:  https://github.com/BaderEddineOuaich/cppcheck-action-other_options-tests
- and this is what the workflow outputs: https://github.com/BaderEddineOuaich/cppcheck-action-other_options-tests/runs/2316871417

if there is a better way to test this please let me know considering i haven't learnt docker yet...
and if testing is a problem, we can use this method until we figure out another better way.
we can make changes to the repos above until it works, then we perform the changes here.

Thanks in advance!

**Comment by deep5050** [comment]
@BaderEddineOuaich  @sthagen 

check: https://github.com/BaderEddineOuaich/cppcheck-action-other_options-tests/runs/2316871417 

```shell
source root not found? Command '['cppcheck', '--enable=all', '--force', '--force', '-i./__tests__/exclude/', '-i./__tests__/exclude/', '--language=c', '--language=c', '--max-ctu-depth=12', '--max-ctu-depth=12', '--std=c++20', '--std=c++20', '--bug-hunting --verbose --std=c11', '--bug-hunting', '--verbose', '--std=c11', '--output-file=./CppCheckReport.txt', '.']' returned non-zero exit status 1.
details:
cppcheck: error: unrecognized command line option: "--bug-hunting --verbose --std=c11".
```
if you look closely: `'--bug-hunting --verbose --std=c11'` is one signle option in the case. along with right `'--bug-hunting'`, `'--verbose'`, `'--std=c11'` seperately

so we are getting the error for the `'--bug-hunting --verbose --std=c11'`, need to remove this. maybe the `split_other_options` needs modification. not sure .

let me know if anyone is interested else I will look into it sometime.

**Comment by sthagen** [comment]
> @BaderEddineOuaich @sthagen
> 
> check: https://github.com/BaderEddineOuaich/cppcheck-action-other_options-tests/runs/2316871417
> 
> ```shell
> source root not found? Command '['cppcheck', '--enable=all', '--force', '--force', '-i./__tests__/exclude/', '-i./__tests__/exclude/', '--language=c', '--language=c', '--max-ctu-depth=12', '--max-ctu-depth=12', '--std=c++20', '--std=c++20', '--bug-hunting --verbose --std=c11', '--bug-hunting', '--verbose', '--std=c11', '--output-file=./CppCheckReport.txt', '.']' returned non-zero exit status 1.
> details:
> cppcheck: error: unrecognized command line option: "--bug-hunting --verbose --std=c11".
> ```
> 
> if you look closely: `'--bug-hunting --verbose --std=c11'` is one signle option in the case. along with right `'--bug-hunting'`, `'--verbose'`, `'--std=c11'` seperately
> 
> so we are getting the error for the `'--bug-hunting --verbose --std=c11'`, need to remove this. maybe the `split_other_options` needs modification. not sure .
> 
> let me know if anyone is interested else I will look into it sometime.

I will take a look (but not before this European evening)

**Comment by BaderEddineOuaich** [comment]
Hello @deep5050 and @sthagen,
this PR works just fine, i mistakenly forgot to remove an extra line when i was applying changes to the other repo which causes that error `cppcheck: error: unrecognized command line option: "--bug-hunting --verbose --std=c11".`
after removing the extra line, the [action](https://github.com/BaderEddineOuaich/cppcheck-action-other_options-tests/runs/2361342144) works like a charm!
sorry guys about this mistake.

**Review by sthagen — COMMENTED** [review]
LGTM

**Comment by deep5050** [comment]
well done guys 👍🏼

**Pull request merged**